scripts/data_collection.py

```diff
@@ -0,0 +1,355 @@
+import csv
+import gc
+import json
+import os
+import random
+import statistics
+import time
+
+import numpy as np
+import torch
+import torchvision.models as models
+from sklearn.neighbors import KDTree
+from torch.profiler import ProfilerActivity, profile, record_function
+from transformers import (
+    AutoConfig,
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    BertConfig,
+    BertForMaskedLM,
+    GPT2ForSequenceClassification,
+    PegasusConfig,
+    PegasusForCausalLM,
+)
+
+torch.set_float32_matmul_precision('high')
+torch.set_default_device('cuda')
+torch.set_default_dtype(torch.float16)
+
+CURR_GPU = "A10G"
+
+
+def timed(fn):
+    start = torch.cuda.Event(enable_timing=True)
+    end = torch.cuda.Event(enable_timing=True)
+    start.record()
+    result = fn()
+    end.record()
+    torch.cuda.synchronize()
+    return result, start.elapsed_time(end) / 1000
+
+
+def percent_error(observed, true):
+    return abs((observed - true) / true) * 100
+
+
+class KDTreeWithValues:
+    def __init__(self, points=None, values=None):
+        self.points = points if points else []
+        self.values = values if values else []
+        if self.points:
+            self.tree = KDTree(self.points)
+        else:
+            self.tree = None
+
+    def add(self, point, value):
+        self.points.append(point)
+        self.values.append(value)
+        self.tree = KDTree(self.points)
+
+    def query(self, point):
+        if self.tree is None:
+            return None, None
+
+        dist, idx = self.tree.query([point], k=1)
+        return dist[0][0], self.values[idx[0][0]]
+
+
+class DataCollectionTreeDB:
+    def __init__(self):
+        self.db = {}
+
+    def add(self, key, point, time):
+        if key not in self.db:
+            self.db[key] = KDTreeWithValues()
+
+        self.db[key].add(point, time)
+
+    def get(self, key, inp):
+        if key not in self.db:
+            # print("New Key")
+            return None
+
+        dist, val = self.db[key].query(inp)
+
+        if dist > 0:
+            # print("Distance too large ", dist)
+            return None
+        return val
+
+
+db = DataCollectionTreeDB()
+
+
+class ShapeProp:
```

Comment:
is it possible to reuse Profiler class for this? Looks like the code is almost exactly same

Reply:
I could reuse some stuff but the logic is a bit different, the data collection needs to run the torch profiler every time it sees something new